keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,168 @@
+
+```go
+type ServiceSpec struct {
+  // topologyKeys is a preference-order list of topology keys.  If backends exist for
+  // index [0], they will always be chosen; only if no backends exist for index [0] will backends for index [1] be considered.
+  // If this field is specified and all indices have no backends, the service has no backends, and connections will fail.  We say these requirements are hard.
+  // In order to experss soft requirement, we may give a special node label key "" as it means "match all nodes".
+  TopologyKeys []string `json:"topologyKeys" protobuf:"bytes,1,opt,name=topologyKeys"`
+}
+```
+
+An example of `Service` with topology keys:
+
+```
+kind: Service
+metadata:
+  name: service-local
+spec:
+  topologyKeys: ["host", "zone"]
+```
+
+
+In our example above, we will firstly try to find the backends in the same host. If no backends match, we will then try the lucky of same zone. If finally we can't find any backends in the same host or same zone, then we say the service has no satisfied backends and connections will fail.
+
+If we configure topologyKeys as `["host", ""]`, we just do the effort to find the backends in the same host and will not fail the connection if no matched backends found.
+
+### New PodLocator resource
+
+As EndpointAddress already contains nodeName field, we can build a service that will precook Pod to its topologies mapping. Then let all interested components(at least kube-proxy and kube-dns and coredns) just watch that precooked object and do necessary mapping internally. Given that we don't know which labels are topology labels, we are going to copy all node labels.
+
+```
+// PodLocator represents information about where a pod exists in arbitrary space.  This is useful for things like
+// being able to reverse-map pod IPs to topology labels, without needing to watch all Pods or all Nodes.
+type PodLocator struct {
+    metav1.TypeMeta
+    // +optional
+    metav1.ObjectMeta
+
+    // NOTE: Fields in this resource must be relatively small and relatively low-churn.
+
+    IPs []PodIPInfo // being added for dual-stack support
+    NodeName string
+    NodeLabels map[string]string
+}
+```
+
+In order to reference PodLocator back to a Pod easily, PodLocator namespace and name would be 1:1 with Pod namespace and name. In other words, PodLocator is a lightweight object which stores Pod location/topology information
+
+### New PodLocator controller
+
+A new PodLocator controller will watch and cache all Pods and Nodes. Then pre-cook pod name to {pod IPs, node name, node labels} mapping. 
+
+When a Pod is added, PodLocator controller will created a new PodLocator object whose namespace and name are 1:1 with Pod namespace and name. Then it will pupolate Pod's IP(s), node name and labels into the new object.
```

Comment:
```suggestion
When a Pod is added, PodLocator controller will created a new PodLocator object whose namespace and name are 1:1 with Pod namespace and name. Then it will populate the Pod's IP(s), node name and labels into the new object.
```